### PR TITLE
Network 1

### PR DESCRIPTION
**Pull request by molbazsa**
This branch is only for experimentation, I just wanted to share the code I'm working on.

### CODE REVIEW

File: LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Layer.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿namespace LinearRegressionBackend.NeuralNetworkPlayground
+{
+    public class Layer
+    {
+
+        public double Weight;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Layer.cs:6]: The Weights should be a Matrix<double> with dimension (l_n-1, l_n).
l_n-1: previous layer neuron number
l_n: current layer neuron number

File: LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Layer.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+            double weight, 
+            double bias, 
+            IActivationFunction activationFunction)
+        {
+            Weight = weight;
+            Bias = bias;
+            ActivationFunction = activationFunction;
+        }
+
+        public (double sum, double activation) Propagate(double input)
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Layer.cs:20]: Propagate input parameter should be a Vector<double> with dimension (l_n-1, 1)
Propagate return type should be a Vector<double> with dimension (l_n, 1)
l_n-1: previous layer neuron number
l_n: current layer neuron number

**Review by SzilvasiPeter — COMMENTED** [review]
It is a good start, now you can scale from single value to vector/matrix value.

**Comment by molbazsa** [comment]
I've implemented the backpropagation for this simple network, now I'll start scaling it to a more complicated network first, then scale it to support more than one training example.

**Comment by molbazsa** [comment]
One thing that's interesting is that I tried to scale the step with the rate of change in the direction of the gradient vector, so that when it's far from the minimum it takes a big step, and when it's close, it takes smaller steps. When I did that, the test failed, but I don't think it's the epochs or the learning rate, because I played around with those and it still failed.
Maybe I calculated the rate of change incorrectly, I used `double rateOfChange = gradient.L2Norm()`, based on this: https://tutorial.math.lamar.edu/classes/calciii/directionalderiv.aspx.
But now I think maybe the whole idea is wrong, because if we take smaller steps whenever we are close to a minimum, we can easily get stuck in a local minimum.

File: LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Sigmoid.cs [hunk 3]
```diff
@@ -11,7 +11,8 @@ public double Activation(double weightedSum)
 
         public double Derivative(double weightedSum)
         {
-            throw new NotImplementedException();
+            return Math.Pow(Math.E, -weightedSum)
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/MLNeuralNetwork/Sigmoid.cs:14]: Sigmoid derivation is correct. However I think the `sigmoid(x) * (1-sigmoid(x))` formula would more computationally efficient, if we store the sigmoid(x) as a variable. (because we don't need to calculate the exp(z) twice)
s = sigmoid(x)
ds = s * (1-s)

File: LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                Layer layer = Layers[i - 1];
+                double z = prop.WeightedSums[i];
+                double a = prop.Activations[i - 1];
+
+                delta *= layer.ActivationFunction.Derivative(z);
+
+                double weightDelta = -learningRate * delta * a;
+                double biasDelta = -learningRate * delta;
+
+                layer.Weight += weightDelta;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs:51]: You need to update the weights in every layer simultaneously. Do not update weights layer by layer.
1. First Calculate the gradient in every layer (dw1, db1, dw2, db2, ...)
2. Second Update the weights and biases (w1=w1-alpha * dw1, b1=b1-alpha * db1, w2=w2-alpha * dw2, b2=b2-aplha * db2)

Why you should update the weights simultaneously? Because you use the weights as you calculate the gradients down the road. If you change the weights then it does not represent the initial state of the neural network anymore.

File: LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        {
+            Propagation result = new();
+
+            result.WeightedSums.Add(0);
+            result.Activations.Add(input);
+
+            foreach (Layer layer in Layers)
+            {
+                (double sum, double activation) = layer.Propagate(input);
+                result.WeightedSums.Add(sum);
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs:25]: Maybe we cache the weightedSum and the Activation into the Layer class instead of the Propagate class

File: LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+            return result;
+        }
+
+        public void Backpropagate(
+            Propagation prop,
+            double expected,
+            double learningRate)
+        {
+            double delta = prop.Output() - expected;
```
Comment (SzilvasiPeter) [inline LinearRegression/LinearRegressionBackend/NeuralNetworkPlayground/Network.cs:38]: I would calculate delta separately from last layer (output layer). delta=(a_L-y) * dsigmoid(z_L), where L is the last layer.
Then I will iterate from the last hidden layer (l=L-1). delta_l=delta_l+1 * w_l+1 * dsigmoid(z_l)
After it is done we can calculate the gradients: dw_l=delta_l * a_l-1, db_l=delta_l

**Review by SzilvasiPeter — COMMENTED** [review]
Good progress, I will draw some clarification about the multiple layer scenario. Basically you have to apply the chain rule religiously. 😄
